stdlib/dynarray.mli

```diff
@@ -0,0 +1,641 @@
+(*   Copyright 2022 Simon Cruanes.                                        *)
+(*   Copyright 2022 Institut National de Recherche en Informatique et     *)
+(*     en Automatique.                                                    *)
+(*                                                                        *)
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+(** Dynamic arrays.
+
+    The {!Array} module provide arrays of fixed length. {!Dynarray}
+    provides arrays whose length can change over time, by adding or
+    removing elements at the end of the array.
+
+    This is typically used to accumulate elements whose number is not
+    known in advance or changes during computation, while also
+    providing fast access to elements at arbitrary positions.
+
+{[
+    let dynarray_of_list li =
+      let arr = Dynarray.create () in
+      List.iter (fun v -> Dynarray.add_last arr v) li;
+      arr
+]}
+
+    The {!Buffer} module provides similar features, but it is
+    specialized for accumulating characters into a dynamically-resized
+    string.
+
+    The {!Stack} module provides a last-in first-out data structure
+    that can be easily implemented on top of dynamic arrays.
+
+    {b Warning.} In their current implementation, the memory layout
+    of dynamic arrays differs from the one of {!Array}s. See the
+    {{!section:memory_layout} Memory Layout} section for more information.
+
+    @since 5.2
+*)
+
+(** {b Unsynchronized accesses} *)
+
+[@@@alert unsynchronized_access
+    "Unsynchronized accesses to dynamic arrays are a programming error."
+]
+
+(**
+   Concurrent accesses to dynamic arrays must be synchronized
+   (for instance with a {!Mutex.t}). Unsynchronized accesses to
+   a dynamic array are a programming error that may lead to an invalid
+   dynamic array state, on which some operations would fail with an
+   [Invalid_argument] exception.
+*)
+
+(** {1:dynarrays Dynamic arrays} *)
+
+type 'a t
+(** A dynamic array containing values of type ['a].
+
+    A dynamic array [a] provides constant-time [get] and [set]
+    operations on indices between [0] and [Dynarray.length a - 1]
+    included. Its {!length} may change over time by adding or removing
+    elements to the end of the array.
+
+    We say that an index into a dynarray [a] is valid if it is in
+    [0 .. length a - 1] and invalid otherwise.
+*)
+
+val create : unit -> 'a t
+(** [create ()] is a new, empty array. *)
+
+val make : int -> 'a -> 'a t
+(** [make n x] is a new array of length [n], filled with [x].
+
+    @raise Invalid_argument if [n < 0] or [n > Sys.max_array_length].
```

Comment:
Currently, the Invalid_argument exception for the upper bound error mentions `Array.make` rather than `Dynarray.make` but that's probably fine.

Reply:
My point of view is that this behavior is more of an internal implementation detail of `Array.init` (which leaks in our specification) than something we should codify in our implementation. So I agree that keeping things as-is is fine.